src/mesh/mesh_tools.C

```diff
@@ -1231,47 +1246,253 @@ void MeshTools::libmesh_assert_valid_amr_interior_parents(const MeshBase & mesh)
+
+  // We want this test to be valid even when called even after nodes
+  // have been added asynchronously but before they're renumbered.
+  //
+  // Plus, some code (looking at you, stitch_meshes) modifies
+  // DofObject ids without keeping max_elem_id()/max_node_id()
+  // consistent, but that's done in a safe way for performance
+  // reasons, so we'll play along and just figure out new max ids
+  // ourselves.
+  dof_id_type parallel_max_node_id = 0;
   for (const auto & node : mesh.node_ptr_range())
+    parallel_max_node_id = std::max<dof_id_type>(parallel_max_node_id,
+                                                 node->id()+1);
+  mesh.comm().max(parallel_max_node_id);
+
+
+  std::vector<bool> node_touched_by_me(parallel_max_node_id, false);
+
+  for (const auto & elem : as_range(mesh.local_elements_begin(),
+                                    mesh.local_elements_end()))
+    {
+      libmesh_assert (elem);
+
+      for (auto & node : elem->node_ref_range())
+        {
+          dof_id_type nodeid = node.id();
+          node_touched_by_me[nodeid] = true;
+        }
+    }
+  std::vector<bool> node_touched_by_anyone(node_touched_by_me);
+  mesh.comm().max(node_touched_by_anyone);
+
+  for (const auto & node : mesh.local_node_ptr_range())
     {
       libmesh_assert(node);
-      libmesh_assert(used_nodes.count(node));
+      dof_id_type nodeid = node->id();
+      libmesh_assert(!node_touched_by_anyone[nodeid] ||
+                     node_touched_by_me[nodeid]);
     }
 }
 
 
 
-void MeshTools::libmesh_assert_valid_constraint_rows (const MeshBase & mesh)
+void libmesh_assert_canonical_node_procids (const MeshBase & mesh)
 {
-  for (auto & row : mesh.get_constraint_rows())
-    {
-      const Node * node = row.first;
-      libmesh_assert(node == mesh.node_ptr(node->id()));
+  for (const auto & elem : mesh.active_element_ptr_range())
+    for (auto & node : elem->node_ref_range())
+      libmesh_assert_equal_to
+        (node.processor_id(),
+         node.choose_processor_id(node.processor_id(),
+                                  elem->processor_id()));
+}
 
-      for (auto & pr : row.second)
+
+
+#ifdef LIBMESH_ENABLE_AMR
+void libmesh_assert_valid_refinement_tree(const MeshBase & mesh)
+{
+  LOG_SCOPE("libmesh_assert_valid_refinement_tree()", "MeshTools");
+
+  for (const auto & elem : mesh.element_ptr_range())
+    {
+      libmesh_assert(elem);
+      if (elem->has_children())
+        for (auto & child : elem->child_ref_range())
+          if (&child != remote_elem)
+            libmesh_assert_equal_to (child.parent(), elem);
+      if (elem->active())
         {
-          const Elem * spline_elem = pr.first.first;
-          libmesh_assert(spline_elem == mesh.elem_ptr(spline_elem->id()));
+          libmesh_assert(!elem->ancestor());
+          libmesh_assert(!elem->subactive());
+        }
+      else if (elem->ancestor())
+        {
+          libmesh_assert(!elem->subactive());
         }
```

Comment:
```suggestion
          libmesh_assert(!elem->subactive());
```

Reply:
I actually like the readability of the redundant version better here.